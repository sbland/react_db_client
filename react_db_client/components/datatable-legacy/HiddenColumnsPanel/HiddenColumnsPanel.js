import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';

let hidePanelTimeout = null;
const HIDETIME = 300;

const HiddenColumnsPanel = ({
  headings,
  hiddenColumnIds,
  handleUnhideColumn,
  showPanelOverride,
}) => {
  const [showPanel, setShowPanel] = useState(showPanelOverride);
  const menuRef = useRef(null);

  function handleClickOutside(event) {
    if (menuRef.current && !menuRef.current.contains(event.target)) {
      hidePanelTimeout = setTimeout(() => setShowPanel(false), HIDETIME);
    } else {
      clearTimeout(hidePanelTimeout);
    }
  }
  useEffect(() => {
    document.addEventListener('mouseover', handleClickOutside);
    return () => {
      // Unbind the event listener on clean up
      clearTimeout(hidePanelTimeout);
      document.removeEventListener('mouseover', handleClickOutside);
    };
  });

  const mapIds = hiddenColumnIds.map((id) => (
    <li key={id}>
      <button type="button" className="button-one" onClick={() => handleUnhideColumn(id)}>
        {headings.find((h) => h.uid === id)?.label || `Missing label for ${id}`}
      </button>
    </li>
  ));

  return (
    <div className="dataTable_hiddenPanel">
      <button
        type="button"
        className="button-one openHiddenButton"
        onClick={() => setShowPanel(!showPanel)}
      >
        Hidden Columns
      </button>
      {showPanel && (
        <ul className="hiddenPanel_hiddenList" ref={menuRef}>
          {showPanel && mapIds}
        </ul>
      )}
    </div>
  );
};

HiddenColumnsPanel.propTypes = {
  headings: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  hiddenColumnIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  handleUnhideColumn: PropTypes.func.isRequired,
  showPanelOverride: PropTypes.bool,
};

HiddenColumnsPanel.defaultProps = {
  showPanelOverride: false,
};

export default HiddenColumnsPanel;
