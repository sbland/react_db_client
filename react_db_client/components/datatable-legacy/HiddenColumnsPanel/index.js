import HiddenColumnsPanel from './HiddenColumnsPanel';

export default HiddenColumnsPanel;
