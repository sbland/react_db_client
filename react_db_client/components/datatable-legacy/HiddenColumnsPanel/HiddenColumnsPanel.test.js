import React from 'react';
import { mount, shallow } from 'enzyme';
import HiddenColumnsPanel from './HiddenColumnsPanel';

const hiddenColumnIds = ['abc', 'def'];
const headings = [
  {uid: 'abc', label: 'ABC'},
  {uid: 'def', label: 'DEF'},
]

describe('Data Table - HiddenColumn Panel', () => {
  test('Snapshot', () => {
    const component = shallow(
      <HiddenColumnsPanel
      headings={headings}
        hiddenColumnIds={hiddenColumnIds}
        setHiddenColumnIds={() => {}}
        showPanelOverride
      />
    );
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });
  const wrapper = mount(
    <HiddenColumnsPanel
    headings={headings}
      hiddenColumnIds={hiddenColumnIds}
      setHiddenColumnIds={() => {}}
      showPanelOverride
    />
  );

  const hiddenList = wrapper.find('.hiddenPanel_hiddenList');
  it('Shows a Row for each hidden item', () => {
    const hiddenCount = hiddenColumnIds.length;
    expect(hiddenList.find('li').length).toEqual(hiddenCount);
  });
});
