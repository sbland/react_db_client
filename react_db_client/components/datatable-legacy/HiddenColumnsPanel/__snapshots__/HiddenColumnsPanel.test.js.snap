// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Table - HiddenColumn Panel Snapshot 1`] = `
"<div className=\\"dataTable_hiddenPanel\\">
  <button type=\\"button\\" className=\\"button-one openHiddenButton\\" onClick={[Function: onClick]}>
    Hidden Columns
  </button>
  <ul className=\\"hiddenPanel_hiddenList\\">
    <li>
      <button type=\\"button\\" className=\\"button-one\\" onClick={[Function: onClick]}>
        abc
      </button>
    </li>
    <li>
      <button type=\\"button\\" className=\\"button-one\\" onClick={[Function: onClick]}>
        def
      </button>
    </li>
  </ul>
</div>"
`;
